File: automedal/agent/phases/strategist.py
"""Strategist phase — curates knowledge.md and writes the next 5 queue entries."""

from __future__ import annotations

from automedal import advisor
from automedal.agent.events import EventSink
from automedal.agent.kernel import RunReport
from automedal.agent.phases._common import run_phase
from automedal.agent.tools import (
    EDIT_FILE, GREP, LIST_DIR, READ_FILE, RECALL, WRITE_FILE,
    make_advisor_tool, make_plan_tool,
)


def _strategist_tools():
    return [READ_FILE, WRITE_FILE, EDIT_FILE, LIST_DIR, GREP, RECALL]


async def run(
    *,
    provider,
    events: EventSink | None,
    exp_id: str,
    iteration: int,
    max_iters: int,
    stagnating: bool | str,
    best_val_loss: float | str,
    pending: int,
    reflective: str,
    ranked: str,
    advisor_note: str = "",
    max_steps: int = 30,
) -> RunReport:
    # Per-phase session dict — scopes plan state to this invocation.
    session: dict = {}

    def _extra_tools(sink):
        extras = [make_plan_tool(session=session, events=sink)]
        if advisor.is_enabled("tool"):
            extras.append(make_advisor_tool(events=sink))
        return extras

    return await run_phase(
        phase="strategist",
        provider=provider,
        tools=_strategist_tools(),
        events=events,
        max_steps=max_steps,
        slots={
            "exp_id": exp_id,
            "iteration": iteration,
            "max_iters": max_iters,
            "stagnating": stagnating,
            "best_val_loss": best_val_loss,
            "pending": pending,
            "reflective": reflective,
            "ranked": ranked,
            "advisor_note": advisor_note,
        },
        extra_tools_factory=_extra_tools,
    )

File: internal/agent/phases/strategist.go
// Strategist phase — curates knowledge.md and writes the next 5 queue entries.
package phases

import (
	"context"

	"github.com/Flameingmoy/automedal/internal/advisor"
	"github.com/Flameingmoy/automedal/internal/agent"
	"github.com/Flameingmoy/automedal/internal/agent/tools"
)

// StrategistArgs is the slot dict for the strategist phase.
type StrategistArgs struct {
	ExpID        string
	Iteration    int
	MaxIters     int
	Stagnating   any
	BestValLoss  any
	Pending      int
	Reflective   string
	Ranked       string
	AdvisorNote  string
	MaxSteps     int
	ConsultFunc  tools.ConsultFunc // wired by cmd layer; nil disables consult_advisor
}

// RunStrategist runs the strategist phase.
func RunStrategist(ctx context.Context, chat agent.ChatStreamFunc, events *agent.EventSink, a StrategistArgs) (agent.RunReport, error) {
	base := []tools.Tool{
		tools.ReadFile, tools.WriteFile, tools.EditFile, tools.ListDir, tools.Grep, tools.Recall,
	}

	// Per-phase session dict — scopes plan state to this invocation.
	session := map[string]any{}

	extraFactory := func(sink *agent.EventSink) []tools.Tool {
		extras := []tools.Tool{tools.MakePlanTool(session, sink)}
		if advisor.IsEnabled("tool") && a.ConsultFunc != nil {
			extras = append(extras, tools.MakeAdvisorTool(a.ConsultFunc, 1))
		}
		return extras
	}

	maxSteps := a.MaxSteps
	if maxSteps <= 0 {
		maxSteps = 30
	}

	return RunPhase(ctx, RunPhaseConfig{
		Phase:    "strategist",
		Chat:     chat,
		Tools:    base,
		Events:   events,
		MaxSteps: maxSteps,
		Slots: map[string]any{
			"exp_id":        a.ExpID,
			"iteration":     a.Iteration,
			"max_iters":     a.MaxIters,
			"stagnating":    a.Stagnating,
			"best_val_loss": a.BestValLoss,
			"pending":       a.Pending,
			"reflective":    a.Reflective,
			"ranked":        a.Ranked,
			"advisor_note":  a.AdvisorNote,
		},
		ExtraToolsFactory: extraFactory,
	})
}
